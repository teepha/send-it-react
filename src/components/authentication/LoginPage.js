import React from "react";
import Spinner from "react-md-spinner";
import { Link, withRouter } from "react-router-dom";
import { connect } from "react-redux";
import { authUserRequest } from "../../actions/userActions";
import { verifyToken } from "../../utils";

export class LoginPage extends React.Component {
  state = {
    email: "",
    password: ""
  };

  componentDidMount() {
    if (verifyToken() !== null) {
      const { user } = this.props;
      if (user.role === "member") {
        this.props.history.push("/user-profile");
      }
      if (user.role === "admin") {
        this.props.history.push("/admin-profile");
      }
    }
  }

  handleLogin = e => {
    e.preventDefault();
    this.props.login(this.state).then(() => {
      const { user } = this.props;
      if (user.role === "member") {
        this.props.history.push("/user-profile");
      }
      if (user.role === "admin") {
        this.props.history.push("/admin-profile");
      }
    });
  };

  handleInputChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    const { loading } = this.props;
    return (
      <div className="main-page">
        <div className="login-wrapper">
          <div className="login-page">
            <Link to="/">
              <i className="fas fa-arrow-left" />
            </Link>
            <h1>LogIn</h1>
            <form onSubmit={this.handleLogin} id="login-form">
              <label>Email address </label>
              <input
                required
                type="email"
                name="email"
                id="email"
                value={this.state.email}
                onChange={this.handleInputChange}
              />
              <br />
              <label>Password </label>
              <input
                required
                name="password"
                type="password"
                id="password"
                value={this.state.password}
                onChange={this.handleInputChange}
              />
              <br />
              <button className="button" type="submit" disabled={loading}>
                {loading ? <Spinner size={18} singleColor="#fff" /> : "Log In"}
              </button>
              <h4 id="error-msg" />
            </form>
            <div className="signup-login-btn">
              <span>Don't have an account?</span>
              <Link to="/register">SignUp</Link>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export const mapStateToProps = ({ user }) => ({
  loading: user.isLoading,
  user: user.userData,
  error: user.userError
});

export const mapDispatchToProps = dispatch => ({
  login: userData => dispatch(authUserRequest(userData))
});

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(LoginPage)
);
