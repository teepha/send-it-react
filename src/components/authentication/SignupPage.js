import React from "react";
import Spinner from "react-md-spinner";
import { Link, withRouter } from "react-router-dom";
import { connect } from "react-redux";
import { authUserRequest } from "../../actions/userActions";
import { verifyToken } from "../../utils";

export class SignupPage extends React.Component {
  state = {
    firstName: "",
    lastName: "",
    phoneNumber: "",
    email: "",
    password: ""
  };

  componentDidMount() {
    if (verifyToken() !== null) {
      const { user } = this.props;
      if (user.role === "member") {
        this.props.history.push("/user-profile");
      }
      if (user.role === "admin") {
        this.props.history.push("/admin-profile");
      }
    }
  }

  handleSignup = e => {
    e.preventDefault();
    this.props.authUserRequest(this.state).then(() => {
      const { user } = this.props;
      if (user.role === "member") {
        this.props.history.push("/user-profile");
      }
      if (user.role === "admin") {
        this.props.history.push("/admin-profile");
      }
    });
  };

  handleInputChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    const { processing } = this.props;
    return (
      <div>
        <div className="main-signup-page">
          <div className="sign-up-wrapper">
            <div className="sign-up-page">
              <Link to="/">
                <i className="fas fa-arrow-left" />
              </Link>
              <h1>SignUp</h1>
              <form onSubmit={this.handleSignup} id="sign-up-form">
                <label>First Name </label>
                <input
                  required
                  pattern="^[A-Za-z]+$"
                  minLength="2"
                  name="firstName"
                  id="first_name"
                  type="text"
                  value={this.state.first_name}
                  onChange={this.handleInputChange}
                />
                <br />
                <label>Last Name </label>
                <input
                  required
                  pattern="^[A-Za-z]+$"
                  minLength="2"
                  name="lastName"
                  id="last_name"
                  type="text"
                  value={this.state.last_name}
                  onChange={this.handleInputChange}
                />
                <br />
                <label>Phone Number </label>
                <input
                  required
                  minLength="11"
                  name="phoneNumber"
                  id="phone_number"
                  type="text"
                  value={this.state.phone_number}
                  onChange={this.handleInputChange}
                />
                <br />
                <label>Email address </label>
                <input
                  required
                  type="email"
                  name="email"
                  id="email"
                  value={this.state.email}
                  onChange={this.handleInputChange}
                />
                <br />
                <label>Password </label>
                <input
                  required
                  minLength="4"
                  name="password"
                  type="password"
                  id="password"
                  value={this.state.password}
                  onChange={this.handleInputChange}
                />
                <br />
                <button className="button" type="submit" disabled={processing}>
                  {processing ? (
                    <Spinner size={18} singleColor="#fff" />
                  ) : (
                    "Sign Up"
                  )}
                </button>
                <h4 id="error-msg" />
                <div className="signup-login-btn">
                  <span>Already have an account?</span>
                  <Link to="/login">LogIn</Link>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ user }) => {
  return {
    processing: user.isLoading,
    user: user.userData,
    error: user.userError
  };
};

const mapDispatchToProps = () => ({
  authUserRequest
});

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps()
  )(SignupPage)
);
