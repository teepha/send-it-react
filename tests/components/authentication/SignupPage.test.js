import React from "react";
import { shallow } from "enzyme";
import { SignupPage } from "../../../src/components/authentication/SignupPage";

jest.mock("../../../src/utils/verifyToken");

describe("LoginPage Component Test", () => {
  const setUp = () => {
    const props = {
      authUserRequest: jest.fn(() => Promise.resolve()),
      processing: false,
      user: {
        id: 1,
        role: "admin"
      },
      error: "",
      history: {
        push: jest.fn()
      }
    };
    const state = {
      firstName: "",
      lastName: "",
      phoneNumber: "",
      email: "",
      password: ""
    };
    return {
      shallowWrapper: shallow(<SignupPage {...props} />),
      state,
      props
    };
  };

  const event = {
    target: {
      firstName: "Seyi",
      lastName: "Adeolu",
      phoneNumber: "08012345678",
      email: "seyi@email.com",
      pasword: "password"
    },
    preventDefault: jest.fn()
  };

  it("render without crashing", () => {
    const { shallowWrapper } = setUp();
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it("should display <Spinner /> component when `isProcessing` is set to true", () => {
    const { shallowWrapper, props } = setUp();
    shallowWrapper.setProps({
      ...props,
      processing: true
    });
    expect(shallowWrapper.find("MDSpinner")).toHaveLength(1);
  });

  it("invokes handleInputChange method", () => {
    const { shallowWrapper } = setUp();
    const handleInputChangeSpy = jest.spyOn(
      shallowWrapper.instance(),
      "handleInputChange"
    );
    shallowWrapper.instance().handleInputChange(event);
    expect(handleInputChangeSpy).toHaveBeenCalled();
  });

  it("invokes handleSignup method and push to admin page", () => {
    const { shallowWrapper, props } = setUp();
    const handleSignupSpy = jest.spyOn(
      shallowWrapper.instance(),
      "handleSignup"
    );

    shallowWrapper.setProps({
      ...props,
      user: {
        ...props.user,
        role: "admin"
      }
    });
    shallowWrapper.instance().handleSignup(event);
    expect(handleSignupSpy).toHaveBeenCalledWith(event);
  });

  it("invokes handleSignup method and push to user page", () => {
    const { shallowWrapper, props } = setUp();
    const handleSignupSpy = jest.spyOn(
      shallowWrapper.instance(),
      "handleSignup"
    );

    shallowWrapper.setProps({
      ...props,
      user: {
        ...props.user,
        role: "member"
      }
    });
    shallowWrapper.instance().handleSignup(event);
    expect(handleSignupSpy).toHaveBeenCalledWith(event);
  });

  it("invokes componentDidMount method", () => {
    const { shallowWrapper, props } = setUp();
    const componentDidMountSpy = jest.spyOn(
      shallowWrapper.instance(),
      "componentDidMount"
    );
    shallowWrapper.instance().componentDidMount();
    expect(componentDidMountSpy).toHaveBeenCalled();

    shallowWrapper.setProps({
      ...props,
      user: {
        ...props.user,
        role: "member"
      }
    });
    shallowWrapper.instance().componentDidMount();
    expect(componentDidMountSpy).toHaveBeenCalled();
  });
});
