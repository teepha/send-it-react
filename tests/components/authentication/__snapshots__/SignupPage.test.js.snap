// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage Component Test render without crashing 1`] = `
<div>
  <div
    className="main-signup-page"
  >
    <div
      className="sign-up-wrapper"
    >
      <div
        className="sign-up-page"
      >
        <Link
          replace={false}
          to="/"
        >
          <i
            className="fas fa-arrow-left"
          />
        </Link>
        <h1>
          SignUp
        </h1>
        <form
          id="sign-up-form"
          onSubmit={[Function]}
        >
          <label>
            First Name 
          </label>
          <input
            id="first_name"
            minLength="2"
            name="firstName"
            onChange={[Function]}
            pattern="^[A-Za-z]+$"
            required={true}
            type="text"
          />
          <br />
          <label>
            Last Name 
          </label>
          <input
            id="last_name"
            minLength="2"
            name="lastName"
            onChange={[Function]}
            pattern="^[A-Za-z]+$"
            required={true}
            type="text"
          />
          <br />
          <label>
            Phone Number 
          </label>
          <input
            id="phone_number"
            minLength="11"
            name="phoneNumber"
            onChange={[Function]}
            required={true}
            type="text"
          />
          <br />
          <label>
            Email address 
          </label>
          <input
            id="email"
            name="email"
            onChange={[Function]}
            required={true}
            type="email"
            value=""
          />
          <br />
          <label>
            Password 
          </label>
          <input
            id="password"
            minLength="4"
            name="password"
            onChange={[Function]}
            required={true}
            type="password"
            value=""
          />
          <br />
          <button
            className="button"
            disabled={false}
            type="submit"
          >
            Sign Up
          </button>
          <h4
            id="error-msg"
          />
          <div
            className="signup-login-btn"
          >
            <span>
              Already have an account?
            </span>
            <Link
              replace={false}
              to="/login"
            >
              LogIn
            </Link>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
